SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialEventTracer.cpp

```diff
@@ -0,0 +1,278 @@
+	}
+	else
+	{
+		CurrentSpanId = Trace_EventTracer_AddSpan(EventTracer, nullptr, 0);
+	}
+
+	auto MessageSrc = StringCast<ANSICHAR>(*SpatialTraceEvent.Message);
+	const ANSICHAR* Message = MessageSrc.Get();
+
+	Trace_Event TraceEvent{ CurrentSpanId, /* unix_timestamp_millis: ignored */ 0, Message, SpatialTraceEvent.Type, nullptr };
+	if (!Trace_EventTracer_ShouldSampleEvent(EventTracer, &TraceEvent))
+	{
+		return {};
+	}
+
+	Trace_EventData* EventData = Trace_EventData_Create();
+
+	for (const auto& Pair : SpatialTraceEvent.Data)
+	{
+		auto KeySrc = StringCast<ANSICHAR>(*Pair.Key);
+		const ANSICHAR* Key = KeySrc.Get();
+		auto ValueSrc = StringCast<ANSICHAR>(*Pair.Value);
+		const ANSICHAR* Value = ValueSrc.Get();
+		Trace_EventData_AddStringFields(EventData, 1, &Key, &Value);
+	}
+
+	// Frame counter
+	{
+		const char* FrameCountStr = "FrameNum";
+		char TmpBuffer[64];
+		FCStringAnsi::Sprintf(TmpBuffer, "%" PRIu64, GFrameCounter);
+		const char* TmpBufferPtr = TmpBuffer;
+		Trace_EventData_AddStringFields(EventData, 1, &FrameCountStr, &TmpBufferPtr);
+	}
+
+	TraceEvent.data = EventData;
+	Trace_EventTracer_AddEvent(EventTracer, &TraceEvent);
+	Trace_EventData_Destroy(EventData);
+
+	return CurrentSpanId;
+}
+
+bool SpatialEventTracer::IsEnabled() const
+{
+	return bEnabled; // Trace_EventTracer_IsEnabled(EventTracer);
+}
+
+void SpatialEventTracer::Enable(const FString& FileName)
+{
+	Trace_EventTracer_Parameters parameters = {};
+	parameters.user_data = this;
+	parameters.callback = &SpatialEventTracer::TraceCallback;
+	EventTracer = Trace_EventTracer_Create(&parameters);
+	Trace_EventTracer_Enable(EventTracer);
+	bEnabled = true;
+
+	UE_LOG(LogSpatialEventTracer, Log, TEXT("Spatial event tracing enabled."));
+
+	// Open a local file
+	FString EventTracePath = FPaths::Combine(FPaths::ProjectSavedDir(), TEXT("EventTracing"));
+	FString AbsLogPath;
+	if (FParse::Value(FCommandLine::Get(), TEXT("eventLogPath="), AbsLogPath, false))
+	{
+		EventTracePath = FPaths::GetPath(AbsLogPath);
+	}
+	const FString FolderPath = EventTracePath;
+
+	const FString FullFileName = FString::Printf(TEXT("EventTrace_%s_%s.trace"), *FileName, *FDateTime::Now().ToString());
+	const FString FilePath = FPaths::Combine(FolderPath, FullFileName);
+
+	IPlatformFile& PlatformFile = FPlatformFileManager::Get().GetPlatformFile();
+	if (PlatformFile.CreateDirectoryTree(*FolderPath))
+	{
+		UE_LOG(LogSpatialEventTracer, Log, TEXT("Capturing trace to %s."), *FilePath);
+		Stream.Reset(Io_CreateFileStream(TCHAR_TO_ANSI(*FilePath), Io_OpenMode::IO_OPEN_MODE_WRITE));
+	}
+}
+
+void SpatialEventTracer::StreamDeleter::operator()(worker::c::Io_Stream* StreamToDestroy) const
+{
+	Io_Stream_Destroy(StreamToDestroy);
+}
+
+void SpatialEventTracer::Disable()
+{
+	UE_LOG(LogSpatialEventTracer, Log, TEXT("Spatial event tracing disabled."));
+
+	Trace_EventTracer_Disable(EventTracer);
+	Trace_EventTracer_Destroy(EventTracer);
+
+	bEnabled = false;
```

Comment:
If this can only be called in the destructor then we don't need to reset to primitive types and pointers. I'd possibly inline this into the destructor to make that clear though.